core/arch/arm/plat-imx/psci.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2016 Freescale Semiconductor, Inc.
+ * All rights reserved.
+ *
+ * Peng Fan <peng.fan@nxp.com>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <console.h>
+#include <drivers/imx_uart.h>
+#include <io.h>
+#include <kernel/generic_boot.h>
+#include <kernel/panic.h>
+#include <kernel/pm_stubs.h>
+#include <mm/core_mmu.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <stdint.h>
+#include <sm/optee_smc.h>
+#include <sm/psci.h>
+#include <tee/entry_std.h>
+#include <tee/entry_fast.h>
+
+static vaddr_t src_base(void)
+{
+       static void *va __data; /* in case it's used before .bss is cleared */
+
+       if (cpu_mmu_enabled()) {
+               if (!va)
+                       va = phys_to_virt(SRC_BASE, MEM_AREA_IO_SEC);
+               return (vaddr_t)va;
+       }
+       return SRC_BASE;
+}
+
+extern paddr_t ns_entry_addrs[CFG_TEE_CORE_NB_CORE];
+int psci_cpu_on(uint32_t core_idx, uint32_t entry,
+		uint32_t context_id __attribute__((unused)))
+{
+	uint32_t val;
+	vaddr_t va = src_base();
+
+	if ((core_idx == 0) || (core_idx >= CFG_TEE_CORE_NB_CORE))
+		return PSCI_RET_INVALID_PARAMETERS;
+
+	/* set secondary cores' NS entry addresses */
+
+	ns_entry_addrs[core_idx] = entry;
+	cache_maintenance_l1(DCACHE_AREA_CLEAN,
+		&ns_entry_addrs[core_idx],
+		sizeof(uint32_t));
+	cache_maintenance_l2(L2CACHE_AREA_CLEAN,
```

Comment:
This call can happen while another core is active in normal world. If there's an active core in normal world it can happen to do an L2 maintenance operation at the same time as this core. 

Now when I read the `cache_maintenance_l2()` function again I see that it's calling `tee_l2cc_mutex_lock()` before doing anything so syncing as actually taken care of provided that normal world has used one of `OPTEE_SMC_L2CC_MUTEX_GET_ADDR`, `OPTEE_SMC_L2CC_MUTEX_SET_ADDR`, or `OPTEE_SMC_L2CC_MUTEX_ENABLE` and uses the spinlock properly.